crates/sui-framework/packages/sui-framework/sources/token.move

```diff
@@ -0,0 +1,731 @@
+        /// against `TokenPolicy.rules` to determine if the request can be
+        /// confirmed.
+        approvals: VecSet<TypeName>,
+    }
+
+    /// Dynamic field key for the `TokenPolicy` to store the `Config` for a
+    /// specific action `Rule`. There can be only one configuration per
+    /// `Rule` per `TokenPolicy`.
+    struct RuleKey<phantom T> has store, copy, drop { is_protected: bool }
+
+    /// Create a new `TokenPolicy` and a matching `TokenPolicyCap`.
+    /// The `TokenPolicy` must then be shared using the `share_policy` method.
+    ///
+    /// `TreasuryCap` guarantees full ownership over the currency, and is unique,
+    /// hence it is safe to use it for authorization.
+    public fun new<T>(
+        _treasury_cap: &TreasuryCap<T>, ctx: &mut TxContext
+    ): (TokenPolicy<T>, TokenPolicyCap<T>) {
+        let policy = TokenPolicy {
+            id: object::new(ctx),
+            spent_balance: balance::zero(),
+            rules: vec_map::empty()
+        };
+
+        let cap = TokenPolicyCap {
+            id: object::new(ctx),
+            for: object::id(&policy)
+        };
+
+        (policy, cap)
+    }
+
+    #[lint_allow(share_owned)]
+    /// Share the `TokenPolicy`. Due to `key`-only restriction, it must be
+    /// shared after initialization.
+    public fun share_policy<T>(policy: TokenPolicy<T>) {
+        transfer::share_object(policy)
+    }
+
+    // === Protected Actions ===
+
+    /// Transfer a `Token` to a `recipient`. Creates an `ActionRequest` for the
+    /// "transfer" action. The `ActionRequest` contains the `recipient` field
+    /// to be used in verification.
+    public fun transfer<T>(
+        t: Token<T>, recipient: address, ctx: &mut TxContext
+    ): ActionRequest<T> {
+        let amount = balance::value(&t.balance);
+        transfer::transfer(t, recipient);
+
+        new_request(
+            transfer_action(),
+            amount,
+            option::some(recipient),
+            option::none(),
+            ctx
+        )
+    }
+
+    /// Spend a `Token` by unwrapping it and storing the `Balance` in the
+    /// `ActionRequest` for the "spend" action. The `ActionRequest` contains
+    /// the `spent_balance` field to be used in verification.
+    ///
+    /// Spend action requires `confirm_request_mut` to be called to confirm the
+    /// request and join the spent balance with the `TokenPolicy.spent_balance`.
+    public fun spend<T>(t: Token<T>, ctx: &mut TxContext): ActionRequest<T> {
+        let Token { id, balance } = t;
+        object::delete(id);
+
+        new_request(
+            spend_action(),
+            balance::value(&balance),
+            option::none(),
+            option::some(balance),
+            ctx
+        )
+    }
+
+    /// Convert `Token` into an open `Coin`. Creates an `ActionRequest` for the
+    /// "to_coin" action.
+    public fun to_coin<T>(
+        t: Token<T>, ctx: &mut TxContext
```

Comment:
I think we can replace `t` with `token` for better naming?

Reply:
Why would you do that?